pytorch_lightning/loops/base.py

```diff
@@ -85,12 +172,18 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
         Returns:
             the output of :attr:`on_run_end` (often outputs collected from each step of the loop)
         """
+        if self.trainer is None:
+            raise MisconfigurationException(f"The {self.__class__.__name__} Loop hasn't been attached to any Trainer.")
+
         if self.skip:
             return self.on_skip()
 
         if self.restarting:
-            self.restore()
-            self.restarting = False
+            if not is_overridden("restore", self, Loop):
+                warning_cache.warn(f"{self.__class__.__name__} Loop doesn't override the restore function.")
+            is_restore_finished = self.restore()
+            if is_restore_finished is None or is_restore_finished:
+                self.restarting = False
```

Comment:
instead of 
```
if restarting
   self.restore()
else:
    self.reset()
```

would it make sense to do:

```
self.reset()
if restarting:
    self.restore()

```
?

Then we don't need the `self._initialize()` in the subclassed loops and we can assume after `reset()` the state is properly initialized and we can restore if appropritae.